fennel/testing/execute_aggregation.py

```diff
@@ -539,29 +584,43 @@ def preprocess_df_discrete_window(
             del_df[FENNEL_ROW_TYPE] = -1
             del_df[ts_field] = del_df[FENNEL_DELETE_TIMESTAMP]
             input_df = pd.concat([input_df, del_df], ignore_index=True)
+
+    # Get the window field from end timestamp, then first cast arrow type and then to frozendict
+    if is_window_key_field:
+        input_df[WINDOW_FIELD_NAME] = cast_col_to_arrow_dtype(
+            input_df[ts_field].apply(
+                lambda x: get_window_from_end_timestamp(x, duration)
+            ),
+            get_window_data_type(),
+        ).apply(lambda x: frozendict(x))
     return input_df
 
 
 def preprocess_df(
```

Comment:
I think we should rename this function to something more meaningful like add_deletes or something

Reply:
changed